### PR TITLE
Make MAINPID= and PIDFile= handling more restrictive (and other stuff)

### PR DESCRIPTION
**Pull request by poettering**
Mostly stuff to make MAINPID= sd_notify() and PIDFile= handling a bit stricter. Plus other minor stuff, some in preparation and some not.

Fixes: #6632.

### CODE REVIEW

File: src/core/dbus-util.c [hunk 1]
```diff
@@ -93,20 +93,20 @@ int bus_set_transient_usec_internal(
...
                 sd_bus_error *error) {
 
-        usec_t v;
+        uint64_t v;
         int r;
 
         assert(p);
 
-        r = sd_bus_message_read(message, "u", &v);
+        r = sd_bus_message_read(message, "t", &v);
```
Comment (yuwata) [inline src/core/dbus-util.c:101]: Oh, thank you...

**Comment by Keensyst** [comment]
If I don't get some money in my wallet today I'm dropping bitcoin, GitHub,
tweeter, Facebook and the rest.

File: src/udev/net/ethtool-util.c [hunk 2]
```diff
@@ -557,7 +557,6 @@ int ethtool_set_glinksettings(int *fd, const char *ifname, struct link_config *l
 
         r = get_glinksettings(fd, &ifr, &u);
         if (r < 0) {
-
```
Comment (yuwata) [inline src/udev/net/ethtool-util.c]: commit title: ehtool -> ethtool

Comment (poettering) [reply src/udev/net/ethtool-util.c]: > commit title: ehtool -> ethtool

pushed a new version fixing that

**Comment by vincentbernat** [comment]
Isn't there a race condition in the checks? Suppose I am forking a new main process (with a new configuration). The old one stays around for cleanup purpose (waiting for old connections to terminate). It signals the new main process with `sd_notify()`. Unfortunately, the new main process dies early. systemd says the PID is not a valid one and MAINPID stays unchanged. The service won't be correctly restarted until the old main process terminates its cleanup.

If there was a way for systemd to acknowledge the MAINPID change somehow, this would fix the problem (old main process would pause new main process until the acknowledgment).

**Comment by poettering** [comment]
We probably should add proper dbus APIs for the various things you can do with sd_notify(), and that would be naturally synchronous, and could return error.  However, I wonder if it would be as attractive to people, given that D-Bus isn't universally loved still...

In the scenario you describe things wouldn't be too bad, after all things would eventually recover. But of course that could be delayed arbitrarily long...

**Pull request merged**